### PR TITLE
curvefs/metaserver: using append instead of update inode

### PR DESCRIPTION
**Pull request by h0hmj**
1. using append instead update inode when s3compact
2. enable s3compact by default

### CODE REVIEW

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 1]
```diff
@@ -295,22 +295,24 @@ int S3CompactWorkQueueImpl::ReadFullChunk(
...
     request.set_poolid(pinfo.poolid());
     request.set_copysetid(pinfo.copysetid());
     request.set_partitionid(pinfo.partitionid());
     request.set_fsid(pinfo.fsid());
-    request.set_inodeid(inode.inodeid());
-    *request.mutable_s3chunkinfomap() = inode.s3chunkinfomap();
-    UpdateInodeResponse response;
-    S3CompactWorkQueueImpl::UpdateInodeClosure done;
+    request.set_inodeid(inodeId);
+    *request.mutable_s3chunkinfoadd() = s3ChunkInfoAdd;
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:308]: `... = std::move(s3ChunkInfoAdd)`

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 2]
```diff
@@ -295,22 +295,24 @@ int S3CompactWorkQueueImpl::ReadFullChunk(
...
+    *request.mutable_s3chunkinforemove() = s3ChunkInfoRemove;
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:309]: ditto

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 3]
```diff
@@ -488,26 +493,28 @@ void S3CompactWorkQueueImpl::CompactChunks(
...
         S3ChunkInfo toAdd;
         toAdd.set_chunkid(newChunkid);
         toAdd.set_compaction(newCompaction);
         toAdd.set_offset(validList.begin()->begin);
         toAdd.set_len(fullChunk.length());
         toAdd.set_size(fullChunk.length());
         toAdd.set_zero(false);
-        *s3chunkVec->Add() = std::move(toAdd);
+        *toAddList.add_s3chunks() = std::move(toAdd);
+        s3ChunkInfoAdd.insert({index, toAddList});
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:507]: `s3ChunkInfoAdd.insert({index, std::move(toAddList)});`

File: curvefs/test/metaserver/mock_s3compactwq_impl.h [hunk 4]
```diff
@@ -46,8 +46,19 @@ class MockS3CompactWorkQueueImpl : public S3CompactWorkQueueImpl {
...
+    MetaStatusCode UpdateInode(
+        CopysetNode* copysetNode, const PartitionInfo& pinfo, uint64_t inodeId,
+        ::google::protobuf::Map<uint64_t, S3ChunkInfoList>&& s3ChunkInfoAdd,
+        ::google::protobuf::Map<uint64_t, S3ChunkInfoList>&&
+            s3ChunkInfoRemove) {
+        return UpdateInode_rvr(copysetNode, pinfo, inodeId, s3ChunkInfoAdd,
+                               s3ChunkInfoRemove);
+    }
+    MOCK_METHOD5(
+        UpdateInode_rvr,
```
Comment (wu-hanqing) [inline curvefs/test/metaserver/mock_s3compactwq_impl.h:58]: what's the `rvr` mean?

**Review by wu-hanqing — COMMENTED** [review]
it seems `MemoryInodeStorage::MergeTwoS3ChunkInfoMap` is unused anymore, if so you shoud remove this function.

File: curvefs/src/metaserver/s3compact_wq_impl.cpp (hunk 1 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp]: fix

File: curvefs/src/metaserver/s3compact_wq_impl.cpp (hunk 2 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp]: fix

File: curvefs/src/metaserver/s3compact_wq_impl.cpp (hunk 3 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp]: fix

**Comment by h0hmj** [comment]
> it seems `MemoryInodeStorage::MergeTwoS3ChunkInfoMap` is unused anymore, if so you shoud remove this function.

yes

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 5]
```diff
@@ -295,22 +295,24 @@ int S3CompactWorkQueueImpl::ReadFullChunk(
...
     request.set_poolid(pinfo.poolid());
     request.set_copysetid(pinfo.copysetid());
     request.set_partitionid(pinfo.partitionid());
     request.set_fsid(pinfo.fsid());
-    request.set_inodeid(inode.inodeid());
-    *request.mutable_s3chunkinfomap() = inode.s3chunkinfomap();
-    UpdateInodeResponse response;
-    S3CompactWorkQueueImpl::UpdateInodeClosure done;
+    request.set_inodeid(inodeId);
+    *request.mutable_s3chunkinfoadd() = std::move(s3ChunkInfoAdd);
```
Comment (ilixiaocui) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:308]: ::google::protobuf::Map<uint64_t, S3ChunkInfoList>&& s3ChunkInfoAdd is rvalue, why here use std::move to convert again?

Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp:308]: Every named object is Lvalue
https://en.cppreference.com/w/cpp/language/value_category
"Even if the variable's type is rvalue reference, the expression consisting of its name is an lvalue expression; "

Comment (ilixiaocui) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp:308]: why just use :google::protobuf::Map<uint64_t, S3ChunkInfoList>& s3ChunkInfoAdd?

Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp:308]: google c++ style guide? const& or pointer*

**Review by ilixiaocui — APPROVED** [review]

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**